common/nexus/outgoing_service_registry.go

```diff
@@ -0,0 +1,422 @@
+	UpdateNamespace(ctx context.Context, request *persistence.UpdateNamespaceRequest) error
+}
+
+// NewOutgoingServiceRegistry creates a new OutgoingServiceRegistry with the given namespace service and configuration.
+func NewOutgoingServiceRegistry(
+	namespaceService NamespaceService,
+	config *OutgoingServiceRegistryConfig,
+) *OutgoingServiceRegistry {
+	return &OutgoingServiceRegistry{
+		namespaceService: namespaceService,
+		config:           config,
+	}
+}
+
+type OutgoingServiceRegistryConfig struct {
+	MaxURLLength    dynamicconfig.IntPropertyFn
+	NameMaxLength   dynamicconfig.IntPropertyFn
+	DefaultPageSize dynamicconfig.IntPropertyFn
+	MaxPageSize     dynamicconfig.IntPropertyFn
+}
+
+func NewOutgoingServiceRegistryConfig(dc *dynamicconfig.Collection) *OutgoingServiceRegistryConfig {
+	return &OutgoingServiceRegistryConfig{
+		MaxURLLength:    dc.GetIntProperty(dynamicconfig.OutgoingServiceURLMaxLength, 1000),
+		NameMaxLength:   dc.GetIntProperty(dynamicconfig.OutgoingServiceNameMaxLength, 200),
+		DefaultPageSize: dc.GetIntProperty(dynamicconfig.OutgoingServiceListDefaultPageSize, 100),
+		MaxPageSize:     dc.GetIntProperty(dynamicconfig.OutgoingServiceListMaxPageSize, 1000),
+	}
+}
+
+var (
+	ErrNamespaceNotSet = status.Errorf(codes.InvalidArgument, "Namespace is not set on request")
+	ErrNameNotSet      = status.Errorf(codes.InvalidArgument, "Name is not set on request")
+	ErrURLNotSet       = status.Errorf(codes.InvalidArgument, "URL is not set on request")
+)
+
+var serviceNameRegex = regexp.MustCompile(`[a-zA-Z_][a-zA-Z0-9_]*`)
+
+func (h *OutgoingServiceRegistry) Get(
+	ctx context.Context,
+	req *operatorservice.GetNexusOutgoingServiceRequest,
+) (*operatorservice.GetNexusOutgoingServiceResponse, error) {
+	if err := validateCommonRequestParams(req); err != nil {
+		return nil, err
+	}
+	response, err := h.namespaceService.GetNamespace(ctx, &persistence.GetNamespaceRequest{
+		Name: req.Namespace,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	service := h.findService(response, req.Name)
+	if service == nil {
+		return nil, status.Errorf(codes.NotFound, "outgoing service %q not found", req.Name)
+	}
+	return &operatorservice.GetNexusOutgoingServiceResponse{
+		Service: service,
+	}, nil
+}
+
+func (h *OutgoingServiceRegistry) Create(
+	ctx context.Context,
+	req *operatorservice.CreateNexusOutgoingServiceRequest,
+) (*operatorservice.CreateNexusOutgoingServiceResponse, error) {
+	if err := h.validateUpsertRequest(req); err != nil {
+		return nil, err
+	}
+	response, err := h.namespaceService.GetNamespace(ctx, &persistence.GetNamespaceRequest{
+		Name: req.GetNamespace(),
+	})
+	if err != nil {
+		return nil, err
+	}
+	ns := response.Namespace
+	i := slices.IndexFunc(ns.OutgoingServices, func(svc *persistencespb.OutgoingService) bool {
+		return svc.Name == req.GetName()
+	})
+	if i >= 0 {
+		return nil, status.Errorf(
+			codes.AlreadyExists,
+			"Outgoing service %q already exists with version %d",
```

Comment:
I actually disagree with this convention when it comes to network errors since they're user-facing, so they should be more readable, similar to logger messages. I think the rationale makes sense for errors that will stay in-memory and are likely to be wrapped. However, I'll make them lowercase.